### PR TITLE
Add moss

### PR DESCRIPTION
**Pull request by AzaleeX**
## Introduction
Addition of blocks / foam mats with breaking time with a hoe add in the BlockBreakInfo

## Hoe add in the BlockBreakInfo

https://github.com/AzaleeX/PocketMine-MP/blob/04eb5ddabc8e3debd8a5945e8a84d797fe56849b/src/block/BlockBreakInfo.php#L72-#L74

## Add Moss Block
https://github.com/AzaleeX/PocketMine-MP/blob/04eb5ddabc8e3debd8a5945e8a84d797fe56849b/src/block/Moss.php#L1-L36

## Add Carpet Moss

https://github.com/AzaleeX/PocketMine-MP/blob/04eb5ddabc8e3debd8a5945e8a84d797fe56849b/src/block/MossCarpet.php#L1-L75

## Tests Screen

image

### CODE REVIEW

File: resources/item_from_string_bc_map.json [hunk 1]
```diff
@@ -1263,6 +1263,7 @@
     "mossy_cobblestone_stairs": "minecraft:mossy_cobblestone_stairs",
     "mossy_stone": "minecraft:mossy_cobblestone",
     "mossy_stone_brick_stairs": "minecraft:mossy_stone_brick_stairs",
+    "moss_block": "minecraft:moss_block",
```
Comment (JavierLeon9966) [inline resources/item_from_string_bc_map.json]: You aren't supposed to modify this file.

File: src/PocketMine.php [hunk 2]
```diff
@@ -17,7 +17,7 @@
  * @link http://www.pocketmine.net/
  *
  *
-*/
+ */
```
Comment (JavierLeon9966) [inline src/PocketMine.php]: Revert the changes from this file.

File: src/block/BlockTypeIds.php [hunk 3]
```diff
@@ -737,8 +737,9 @@ private function __construct(){
 	public const BIG_DRIPLEAF_HEAD = 10707;
 	public const BIG_DRIPLEAF_STEM = 10708;
 	public const PINK_PETALS = 10709;
-
 	public const FIRST_UNUSED_BLOCK_ID = 10710;
+	public const MOSS_BLOCK = 10711;
```
Comment (JavierLeon9966) [inline src/block/BlockTypeIds.php]: New block type IDs must not exceed the `FIRST_UNUSED_BLOCK_ID` value.

File: src/block/MossCarpet.php [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+	}
+
+	public function getFlameEncouragement() : int{
+		return 30;
+	}
+
+	public function getFlammability() : int{
+		return 20;
+	}
+
```
Comment (JavierLeon9966) [inline src/block/MossCarpet.php]:
```suggestion
```

File: src/block/Moss.php [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+declare(strict_types=1);
+
+namespace pocketmine\block;
+
+use pocketmine\item\Item;
+
+class Moss extends Opaque{
+
+	public function getDropsForCompatibleTool(Item $item) : array{
```
Comment (JavierLeon9966) [inline src/block/Moss.php]: Why do we need this function and class? You can obtain moss with your hand. You can replace it with `new Opaque(...)`.

File: src/block/MossCarpet.php [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+use pocketmine\player\Player;
+use pocketmine\world\BlockTransaction;
+
+class MossCarpet extends Flowable{
+
+	public function isSolid() : bool{
+		return true;
+	}
+
+	public function getDropsForCompatibleTool(Item $item) : array{
```
Comment (JavierLeon9966) [inline src/block/MossCarpet.php]: Same here.

File: src/block/VanillaBlocks.php [hunk 7]
```diff
@@ -829,6 +831,10 @@ public function getBreakTime(Item $item) : float{
 		self::register("cobblestone_stairs", new Stair(new BID(Ids::COBBLESTONE_STAIRS), "Cobblestone Stairs", $cobblestoneBreakInfo));
 		self::register("mossy_cobblestone_stairs", new Stair(new BID(Ids::MOSSY_COBBLESTONE_STAIRS), "Mossy Cobblestone Stairs", $cobblestoneBreakInfo));
 
+		$mossBreakInfo = new Info(BreakInfo::pickaxe(0.1));
```
Comment (JavierLeon9966) [inline src/block/VanillaBlocks.php]:
```suggestion
		$mossBreakInfo = new Info(BreakInfo::hoe(0.1));
```

**Review by JavierLeon9966 — CHANGES_REQUESTED** [review]

**Comment by JavierLeon9966** [comment]
One thing left are Moss blocks spreading when interacted with a fertilizer and replacing some blocks like stone.

File: src/block/VanillaBlocks.php (hunk 7 above)
Comment (AzaleeX) [reply src/block/VanillaBlocks.php:834]: Why pickaxe ?

File: src/block/BlockTypeIds.php (hunk 3 above)
Comment (AzaleeX) [reply src/block/BlockTypeIds.php]: So what am I supposed to do ?

Comment (JavierLeon9966) [reply src/block/BlockTypeIds.php]: Shift first unused block type ID after Moss block and Moss carpet.

File: src/block/VanillaBlocks.php (hunk 7 above)
Comment (JavierLeon9966) [reply src/block/VanillaBlocks.php]: You wrote it as pickaxe instead of hoe.

Comment (AzaleeX) [reply src/block/VanillaBlocks.php]: I'm sorry I created a function on purpose hoe but I don't use it :/

File: src/block/BlockTypeIds.php (hunk 3 above)
Comment (AzaleeX) [reply src/block/BlockTypeIds.php]:
```php
public const FIRST_UNUSED_BLOCK_ID = 10713;
```

Comment (AzaleeX) [reply src/block/BlockTypeIds.php]: this must be correct ?

Comment (JavierLeon9966) [reply src/block/BlockTypeIds.php]: Yes but you would be leaving a number gap between Moss block and pink petals.

File: src/block/Moss.php (hunk 5 above)
Comment (AzaleeX) [reply src/block/Moss.php]: I find that for me having the blocks in a file allows the developer to know which block is made and others don't like me who wants to implement all the blocks that pocketmine does not support

File: src/block/Moss.php [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+ *
+ */
+
+declare(strict_types=1);
+
+namespace pocketmine\block;
+
+use pocketmine\item\Item;
+
+class Moss extends Opaque{
```
Comment (JavierLeon9966) [inline src/block/Moss.php]: This class isn't needed. You can use Opaque instead.

File: src/PocketMine.php [hunk 9]
```diff
@@ -36,18 +36,40 @@
 	use pocketmine\utils\Utils;
 	use pocketmine\wizard\SetupWizard;
 	use Symfony\Component\Filesystem\Path;
+	use function count;
```
Comment (JavierLeon9966) [inline src/PocketMine.php]: You still didn't revert the changes from all this file.

**Review by JavierLeon9966 — CHANGES_REQUESTED** [review]

**Comment by AzaleeX** [comment]
> One thing left are Moss blocks spreading when interacted with a fertilizer and replacing some blocks like stone.

I'm thinking of implementing this tomorrow because it's late at home and tomorrow I'm working 👍

File: src/block/Moss.php (hunk 8 above)
Comment (AzaleeX) [reply src/block/Moss.php]: I find that for me having the blocks in a file allows the developer to know which block is made and others don't like me who wants to implement all the blocks that pocketmine does not support

Comment (ShockedPlot7560) [reply src/block/Moss.php]: However, declaring an empty class does nothing but add a bunch of useless files.

Comment (AzaleeX) [reply src/block/Moss.php]: yes it's not wrong but it would be a way that might be useless for you to know which block is already implemented in pocketmine but hey for that we have the VanillaItems just to do a Ctrl + F to search for the block

File: src/PocketMine.php [hunk 10]
```diff
@@ -138,9 +138,9 @@ function check_platform_dependencies(){
 		$wantedVersionLock = "0.3";
 		$wantedVersionMin = "$wantedVersionLock.0";
 		if($chunkutils2_version !== false && (
-			version_compare($chunkutils2_version, $wantedVersionMin) < 0 ||
-			preg_match("/^" . preg_quote($wantedVersionLock, "/") . "\.\d+(?:-dev)?$/", $chunkutils2_version) === 0 //lock in at ^0.2, optionally at a patch release
-		)){
+				version_compare($chunkutils2_version, $wantedVersionMin) < 0 ||
```
Comment (JavierLeon9966) [inline src/PocketMine.php]: Revert all the changes from this file. It's unrelated from the PR.

**Review by JavierLeon9966 — CHANGES_REQUESTED** [review]

File: src/PocketMine.php (hunk 10 above)
Comment (AzaleeX) [reply src/PocketMine.php]: https://github.com/pmmp/PocketMine-MP/blob/stable/src/PocketMine.php#L141

Comment (AzaleeX) [reply src/PocketMine.php]: this is basic in the Pocketmine.php

File: src/block/Moss.php (hunk 8 above)
Comment (JavierLeon9966) [reply src/block/Moss.php]: The class will be added back anyways once the spread feature gets implemented.

Comment (AzaleeX) [reply src/block/Moss.php]: OK 👌

File: src/PocketMine.php (hunk 10 above)
Comment (JavierLeon9966) [reply src/PocketMine.php]: What do you mean?

Comment (AzaleeX) [reply src/PocketMine.php]: this is in the stable branch there is no difference

Comment (JavierLeon9966) [reply src/PocketMine.php]: There are differences because this is in the "Files changed" section.

**Review by JavierLeon9966 — CHANGES_REQUESTED** [review]
One final thing is this must target the `minor-next` branch.

File: src/PocketMine.php (hunk 10 above)
Comment (dktapps) [reply src/PocketMine.php]: These formatting changes must be removed

**Review by dktapps — CHANGES_REQUESTED** [review]
The bonemeal behaviour of these blocks is missing

File: src/PocketMine.php (hunk 10 above)
Comment (AzaleeX) [reply src/PocketMine.php]: > These formatting changes must be removed

this was done in the commits

File: src/PocketMine.php [hunk 11]
```diff
@@ -355,4 +355,4 @@ function server(){
 	}
 
 	\pocketmine\server();
-}
+}
```
Comment (dktapps) [inline src/PocketMine.php:358]: Unrelated formatting change

**Review by dktapps — CHANGES_REQUESTED** [review]
As I said previously, the PR is still missing the bonemeal logic.

File: src/block/MossCarpet.php [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+
+	/**
+	 * @return AxisAlignedBB[]
+	 */
+	protected function recalculateCollisionBoxes() : array{
+		return [AxisAlignedBB::one()->trim(Facing::UP, 15 / 16)];
+	}
+
+	public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Block $blockClicked, int $face, Vector3 $clickVector, ?Player $player = null) : bool{
+		$down = $this->getSide(Facing::DOWN);
```
Comment (ShockedPlot7560) [inline src/block/MossCarpet.php]: Can be merged with your condition

File: src/block/MossCarpet.php [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+		if($down->getTypeId() !== BlockTypeIds::AIR){
```
Comment (ShockedPlot7560) [inline src/block/MossCarpet.php]: Can be moove with the `StaticSupportTrait` trait to use the same placement logic into `onNearbyBlockChange`

File: src/block/MossCarpet.php [hunk 14]
```diff
@@ -44,18 +46,14 @@ protected function recalculateCollisionBoxes() : array{
...
 
 	public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Block $blockClicked, int $face, Vector3 $clickVector, ?Player $player = null) : bool{
-		$down = $this->getSide(Facing::DOWN);
-		if($down->getTypeId() !== BlockTypeIds::AIR){
+		if($this->getSide(Facing::DOWN)->getTypeId() !== BlockTypeIds::AIR){
 			return parent::place($tx, $item, $blockReplace, $blockClicked, $face, $clickVector, $player);
 		}
 
 		return false;
 	}
```
Comment (dktapps) [inline src/block/MossCarpet.php:54]: This whole function can be deleted. StaticSupportTrait will do this anyway.
